### ISSUE TITLE
slimserver fails to communicate with Spotify on Raspberry Pi 4

### ISSUE THREAD

**Issue by mfrischknecht** [NixOS/nixpkgs#109270]
**Describe the bug**
After fixing the problems in #108617 (cf [this branch](https://github.com/NixOS/nixpkgs/compare/master...mfrischknecht:slimserver-Net-HTTPS-NB?expand=1)), the Spotify plugin for slimserver fails to run the Spotify helper binaries it ships with.

**To Reproduce**
1. Set up NixOS for a Raspberry PI 4 using the [images](https://hydra.nixos.org/job/nixos/trunk-combined/nixos.sd_image_raspberrypi4.aarch64-linux) indicated by the [nixos wiki](https://nixos.wiki/wiki/NixOS_on_ARM/Raspberry_Pi_4).
2. Apply the patch outlined in #108617 (cf [this branch](https://github.com/NixOS/nixpkgs/compare/master...mfrischknecht:slimserver-Net-HTTPS-NB?expand=1))
3. Set up `slimserver` (`services.slimserver.enable = true`).
4. After startup, `slimserver` will log the following error message (visible in `journalctl -u slimserver`):
    ```
   Plugins::Spotty::Helper::get (51) Didn't find Spotty helper application!
   Plugins::Spotty::Helper::get (52) Last error:
   sh: /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/arm-linux/spotty-hf: No such file or directory
   ```

   Additionally, in the "Plugins" tab of the settings page, the following text is visible:
   ```
   There has been a problem running the Spotty helper application. You might want to give the additional "Spotty Helper Application for ARMv8/aarch64 architecture" plugin a try.
   Operating system: Linux / aarch64-linux

   Helper Applications Folder:
   /nix/store/iif6k1ldrzlncbwf9fm89xgcgw1ss59z-lame-3.100/bin
   /nix/store/zpiq0sj3nlqz6pkq88f4vqixs5xwzgqp-flac-1.3.3-bin/bin
   /nix/store/ad4h1d0l2z1v05q0pvpsagqmqz6qd24d-faad2-2.10.0/bin
   /nix/store/6rr1y8q5zsy185ahabjzp81aqpfw0f39-sox-14.4.2/bin
   /nix/store/6f2a3df8l8z563ad3dib5bjpj9c28hyb-monkeys-audio-3.99-u4-b5/bin
   /nix/store/x2x9qzggq2vwyl0jn8kalcnqfiifisn8-wavpack-5.3.0/bin
   /nix/store/mc7s5nla8y7nr8x5j1yz1kmq8xh1bhcg-coreutils-8.32/bin
   /nix/store/5xiyazii9004aipigb5bp6rm6aqi4mak-findutils-4.7.0/bin
   /nix/store/qqa9c02q397zk1mqi2n8rd93ksm8xvh3-gnugrep-3.6/bin
   /nix/store/vjhx2pa94y87r50mfmkasf6cmrgff63y-gnused-4.8/bin
   /nix/store/7l8g3cahr0yiszwl3ca8aic59b8ap02v-systemd-247.2/bin
   /nix/store/7l8g3cahr0yiszwl3ca8aic59b8ap02v-systemd-247.2/sbin
   /usr/bin
   /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/aarch64-linux
   /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin
   /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/arm-linux
   sh: ldd: command not found
   ```

On x64, I was able to set slimserver up accordingly and use the Spotify plugin without running into any issues.
When trying to run the correct binary directly from a shell, bash reports the same error:

```
[nixos@nixos:~/source/nix/nixpkgs]$ /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/aarch64-linux/spotty 
-bash: /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/aarch64-linux/spotty: No such file or directory
```

OTOH, this works on x64:
```
λ ~/ /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/i386-linux/spotty
spotty v0.35.0
error: Required option 'name' missing
Usage: /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/i386-linux/spotty [options]

[... and so on; I'll omit the help message here]
```

**Expected behavior**
The `spotty` helper application is runnable and the Spotify plugin for slimserver can be used.

**Notify maintainers**

@phile314

**Metadata**

```
 - system: `"aarch64-linux"`
 - host os: `Linux 4.19.118, NixOS, 21.03pre260232.733e537a8ad (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-21.03pre260232.733e537a8ad"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
- pkgs.slimserver
# a list of nixos modules affected by the problem
module:
- services.slimserver
```

**Comment by mfrischknecht** [comment]
While looking for explanations for this error, I found [this forum post](https://www.raspberrypi.org/forums/viewtopic.php?p=498535#p498535), stating that most likely, a missing program interpreter might be the culprit.

And indeed:

```
[nix-shell:~/source/nix/nixpkgs]$ readelf -a /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/aarch64-linux/spotty | fgrep -i requesting
      [Requesting program interpreter: /lib/ld-linux-aarch64.so.1]

[nix-shell:~/source/nix/nixpkgs]$ readelf -a "$(which ls)" | fgrep -i requesting
      [Requesting program interpreter: /nix/store/4pc5akylpxdi0azn6mcqcg0qaj4wz88r-glibc-2.32-10/lib/ld-linux-aarch64.so.1]
```

At the same time, the helper binary for i386 doesn't seem to contain such a path (i.e. there's no output):
```
[nix-shell:~]$ readelf -a /var/lib/slimserver/cache/InstalledPlugins/Plugins/Spotty/Bin/i386-linux/spotty | fgrep -i requesting

[nix-shell:~]$ 
```

I'm not sure how to proceed from here, though. Is there a way to set up an environment so that binaries that weren't built through Nix itself can still run? Patching them in a nix derivation would seem complicated to me, since they are shipped by the Spotify plugin and not through the slimserver package itself...

**Comment by mfrischknecht** [comment]
Reading the [GitHub readme](https://github.com/michaelherger/Spotty-Plugin) of the Spotty plugin, I discovered that the `spotty` helper application is maintained in a separate [repository](https://github.com/michaelherger/spotty).

The `Bin` directory in the plugin's original repository also contains [some notes](https://github.com/michaelherger/Spotty-Plugin/blob/master/Bin/README.md) on how these helper binaries are supposed to be built on different architectures.

Since the text from the plugin settings page (cf. the original issue description) lists nix store paths before anything else, I'll try to write a new nix derivation expression that builds such a binary on the current architecture and can be included as a build input in the `slimserver` derivation.

If I'm successful, I hope this would lead the server to try executing this binary first, effectively replacing the binaries distributed by the plugin itself. If so, my suggestion would be to introduce a new option to toggle this dependency.

**Comment by phile314** [comment]
Ah,yes, this is going to be problematic. As far as I remember, these plugins are installed using slimserver and are completely outside of the control of nixpkgs, right?

I think your proposed solution sounds like a good approach.

**Comment by mfrischknecht** [comment]
Hi @phile314,

thanks for your response!

Unfortunately, the experiments on the `spotty` helper application I stared yesterday weren't really successful:

The Cargo project of `spotty` uses Cargo's `replace` feature to [switch out](https://github.com/michaelherger/spotty/blob/master/Cargo.lock#L1222) the dependency of `rust-crypto` with a specific feature branch (something related to AES with AVX2 that was [discussed here in librespot](https://github.com/librespot-org/librespot/issues/128)). 

Unfortunately, this clashes with #30742 as `cargo vendor` seemingly still has no proper support for `replace` dependencies, which means building the app simply through `buildRustPackage` fails.

And since `spotty` uses a [custom version of librespot](https://github.com/michaelherger/spotty/blob/97b7ca3cec9999d9131a73935c08eda74b45a9d3/Cargo.lock#L568), I'm currently not sure if I can simply remove the replace definitions or not (I can't find the respective entry in the fork's [Cargo.lock](https://github.com/michaelherger/librespot/blob/8c2bd4e0e2323ee1f835d28a6aa9ede700e7497c/Cargo.lock)). I'll have to investigate further.

**Comment by mfrischknecht** [comment]
I was able to resolve the build problems I mentioned above: `spotty` built fine through Nix once I removed the patched version of `rust-crypto` from the project's dependencies. I pushed the nix code I have running so far to a [branch on my nixpkgs fork](https://github.com/mfrischknecht/nixpkgs/compare/slimserver-raspi...mfrischknecht:spotty?expand=1).

Now, I'm not sure how to proceed. First of all, removing the custom package generates quite a couple of changes to the project's [Cargo.lock](https://github.com/mfrischknecht/nixpkgs/blob/599552be1efbd783eed743b91fecfda159222c48/pkgs/tools/audio/spotty/use-cargo-rust-crypto.patch) file. Also, since the original discussion on the custom `rust-crypto` branch used revolved around [issues on Intel Westmere CPUs](https://github.com/plietar/librespot/issues/155), the resulting binary will most likely not work properly on those. At the same time, since `spotty` uses a customized fork of `librespot`, simply updating that dependency also isn't enough to avoid the `rust-crypto` fork, although the mainline library seemingly [has moved on to another crypto library in 2018](https://github.com/librespot-org/librespot/pull/239).

On my branch, I just added it as another entry under `pkgs/tools/audio`, but given the above, that might not be the best idea. Would it be a better option to add it as a separate nix file to `pkgs/servers/slimserver` and just importing and adding it to `wrapProgram` using `optional`?

...or should I better just maintain this in a separate repository and pull it in via `pkgs.slimserver.overrideAttrs` and `services.slimserver.package`? If so, can you give me a hint on how I could influence the `wrapProgram` call via `overrideAttrs` (I've never done that so far)?

**Comment by mfrischknecht** [comment]
I just realized I failed to mention the most important thing: With the `spotty` helper binary building, the Spotify plugin for `slimserver` actually works. I simply had to add it to the [wrapProgram](https://github.com/mfrischknecht/nixpkgs/blob/599552be1efbd783eed743b91fecfda159222c48/pkgs/servers/slimserver/default.nix#L92) call, and the plugin picked up the right binary.

![Screenshot_2021-01-15 Logitech Media Server](https://user-images.githubusercontent.com/7261570/104724481-02632c80-5731-11eb-90c4-76e7bfc26749.png)

**Comment by phile314** [comment]
I don't know anything about the rust part, but I guess that could be problematic.

I see two options:
1. Add the spotty binary to the path of the systemd service (`systemd.services.slimserver.path`)
2. Add an extra-parameter to the slimserver package derivation to splice in extra binaries to the `wrapProgram` call

I think option 1 is cleaner. I think that way you can also keep spotty in your private branch or try to get it merged into nixpkgs proper, whichever you prefer.

**Comment by mfrischknecht** [comment]
Thank you very much for your hint on `systemd.services.slimserver.path`!

[I tested it here](https://github.com/mfrischknecht/nixos-slimserver-raspi4/blob/9633768f2db9098d22b7219801494c403a442878/services/slimserver/default.nix#L7), and the `spotty` plugin for `slimserver` indeed picks up the correct helper binary if it's provided in this way, making further changes to the `slimserver` derivation unnecessary.

Since my nix derivation for `spotty` currently won't work on every platform (see my comment above), I decided to just pull it in through a custom nix file in my own config (at least until #30742 is fixed). If anyone runs into the same issues, they're welcome to just copy the [spotty.nix](https://github.com/mfrischknecht/nixos-slimserver-raspi4/blob/9633768f2db9098d22b7219801494c403a442878/services/slimserver/spotty.nix) and [use-cargo-rust-crypto.patch](https://github.com/mfrischknecht/nixos-slimserver-raspi4/blob/9633768f2db9098d22b7219801494c403a442878/services/slimserver/use-cargo-rust-crypto.patch) files from my configuration.

Since this setup works for my device (the only issue being some long rust compilation times), I'll close this issue. Thanks again for your help, @phile314! :-)
